uclasm/matching/local_cost_bound/edgewise.py

```diff
@@ -533,44 +533,30 @@ def add_time_costs(smp, candidates, local_costs):
-                    if cand2_idx_idx >= len(cand2_sorted_idxs):
-                        cand2_time = None
-                        break
-                    cand2_idx = cand2_sorted_idxs[cand2_idx_idx]
-                    cand2_time = cand2_times[cand2_idx]
-                if cand2_time is not None:
-                    if 'maxValue' in time_constraint:
-                        if cand2_time - cand1_time > max_timedelta:
-                            continue
-                    cand1_nonnat_costs[cand1_idx] = 0
-                    if 'maxValue' in time_constraint:
-                        cand2_nonnat_costs[cand2_idx] = 0
-                        temp_cand2_idx_idx = max(last_valid_cand2_idx_idx, cand2_idx_idx) + 1
-                        if temp_cand2_idx_idx < len(cand2_sorted_idxs):
-                            next_cand2_idx = cand2_sorted_idxs[temp_cand2_idx_idx]
-                            next_cand2_time = cand2_times[next_cand2_idx]
-                            while next_cand2_time - cand1_time <= max_timedelta:
-                                cand2_nonnat_costs[next_cand2_idx] = 0
-                                last_valid_cand2_idx_idx = temp_cand2_idx_idx
-                                temp_cand2_idx_idx += 1
-                                if temp_cand2_idx_idx < len(cand2_sorted_idxs):
-                                    next_cand2_idx = cand2_sorted_idxs[temp_cand2_idx_idx]
-                                    next_cand2_time = cand2_times[next_cand2_idx]
-                                else:
-                                    break
-                    elif cand1_idx_idx == 0:
-                        # With only minimum timedelta, all future cand2_times are valid
-                        cand2_nonnat_costs[cand2_sorted_idxs[cand2_idx_idx:]] = 0
+            # There are a lot of redundant times, only check unique times
+            uniq_cand1_times = np.unique(cand1_times)
+
+            # For each initial event time
+            for cand1_time in uniq_cand1_times:
+                if 'minValue' in time_constraint:
+                    min_idx = np.searchsorted(cand2_times, cand1_time + min_timedelta,
+                                              sorter=cand2_sorted_idxs)
                 else:
-                    break
+                    min_idx = 0
+                if 'maxValue' in time_constraint:
+                    max_idx = np.searchsorted(cand2_times, cand1_time + max_timedelta,
+                                              'right', sorter=cand2_sorted_idxs)
+                else:
+                    max_idx = len(cand2_times)
+                
+                cand1s = (cand1_times == cand1_time).nonzero()[0]
+
+                # If there are cand2 vertices in desired time range,
+                # set associated costs to 0.
+                if min_idx < max_idx:
+                    cand1_nonnat_costs[cand1s] = 0
+                    cand2_nonnat_costs[cand2_sorted_idxs[min_idx:max_idx+1]] = 0
```

Comment:
max_idx is the index of the smallest value larger than cand1_time + max_timedelta, so it should not be included. So the +1 should be removed.

Reply:
Fixed.